control-base/src/main/java/fi/nls/oskari/control/feature/GetWFSFeaturesHandler.java

```diff
@@ -0,0 +1,236 @@
+            String requestSrsName, String targetSrsName) throws ActionException {
+        CoordinateReferenceSystem requestCRS;
+        CoordinateReferenceSystem targetCRS;
+        try {
+            requestCRS = CRS.decode(requestSrsName, true);
+            targetCRS = CRS.decode(targetSrsName, true);
+        } catch (Exception e) {
+            throw new ActionException(ERR_CRS_DECODE_FAIL);
+        }
+
+        if (!isWithin(targetCRS, bbox)) {
+            throw new ActionParamsException(ERR_BBOX_OUT_OF_CRS);
+        }
+
+        boolean shouldTransform = !CRS.equalsIgnoreMetadata(requestCRS, targetCRS);
+
+        MathTransform transformFromRequested;
+        MathTransform transformToRequested;
+        Envelope requestBbox;
+        if (shouldTransform) {
+            try {
+                transformFromRequested = CRS.findMathTransform(requestCRS, targetCRS, true);
+                transformToRequested = CRS.findMathTransform(targetCRS, requestCRS, true);
+            } catch (Exception e) {
+                throw new ActionException(ERR_TRANSFORM_FIND_FAIL);
+            }
+            try {
+                requestBbox = JTS.transform(bbox, transformToRequested);
+            } catch (Exception e) {
+                throw new ActionException(ERR_REPOJECTION_FAIL);
+            }
+        } else {
+            transformFromRequested = null;
+            transformToRequested = null;
+            requestBbox = bbox;
+        }
+
+        SimpleFeatureCollection features = getFeatures(layer, requestSrsName, requestBbox);
+        if (shouldTransform) {
+            try {
+                features = transform(features, targetCRS, transformFromRequested);
+            } catch (Exception e) {
+                throw new ActionException(ERR_REPOJECTION_FAIL);
+            }
+        }
+        return features;
+    }
+
+    private boolean isWithin(CoordinateReferenceSystem targetCRS, Envelope bbox) {
+        org.opengis.geometry.Envelope targetCRSEnvelope = CRS.getEnvelope(targetCRS);
+        DirectPosition lc = targetCRSEnvelope.getLowerCorner();
+        if (bbox.getMinX() < lc.getOrdinate(0)) {
+            return false;
+        }
+        if (bbox.getMinY() < lc.getOrdinate(1)) {
+            return false;
+        }
+        DirectPosition uc = targetCRSEnvelope.getUpperCorner();
+        if (bbox.getMaxX() > uc.getOrdinate(0)) {
+            return false;
+        }
+        if (bbox.getMaxY() > uc.getOrdinate(1)) {
+            return false;
+        }
+        return true;
+    }
+
+    private SimpleFeatureCollection getFeatures(OskariLayer layer, String srsName, Envelope bbox) {
+        String endPoint = layer.getUrl();
+        String typeName = layer.getName();
+        String user = layer.getUsername();
+        String pass = layer.getPassword();
+        // TODO: Figure out the maxFeatures from the layer 
+        int maxFeatures = 10000;
```

Comment:
This is stored per layer in the old wfs-tables. Perhaps this could be stored in layer.getAttributes()?

Reply:
Yes, that would be nice.